pkg-r/R/contents_shinychat.R

```diff
@@ -1,14 +1,296 @@
+  # Fall back to html or markdown
+  html <- ellmer::contents_html(content)
+  if (!is.null(html)) {
+    shiny::HTML(html)
+  } else {
+    ellmer::contents_markdown(content)
+  }
+}
+
+S7::method(contents_shinychat, ellmer::ContentText) <- function(content) {
+  content@text
+}
+
+S7::method(contents_shinychat, ellmer::ContentToolRequest) <- function(
+  content
+) {
+  if (opt_shinychat_tool_display() == "none") {
+    return(NULL)
+  }
+
+  props <- list(
+    "request-id" = content@id,
+    name = content@name,
+    arguments = jsonlite::toJSON(content@arguments, auto_unbox = TRUE),
+    intent = content@arguments$.intent
+  )
+
+  tool <- content@tool
+  if (!is.null(tool)) {
+    props$title <- tool@annotations$title
+  }
+
+  tagList(
+    htmltools::tag("shiny-tool-request", props),
+    chat_deps()
+  )
+}
+
+S7::method(contents_shinychat, ellmer::ContentToolResult) <- function(content) {
+  if (opt_shinychat_tool_display() == "none") {
+    return(NULL)
+  }
+
+  if (is.null(content@request)) {
+    cli::cli_abort(
+      "`ContentToolResult` objects must have an associated `@request` property."
+    )
+  }
+
+  content@extra$display <- ensure_content_display(content)
+
+  # Prepare base props
+  props <- list(
+    request_id = content@request@id,
+    request_call = "",
+    name = content@request@name,
+    status = if (tool_errored(content)) "error" else "success",
+    show_request = if (!isFALSE(content@extra$display$show_request)) NA,
+    intent = content@request@arguments$.intent,
+    expanded = if (isTRUE(content@extra$display$open)) NA
+  )
+
+  icon_deps <- NULL
+  tool <- content@request@tool
+
+  # Let tool results override global tool properties
+  if (!is.null(content@extra$display$title)) {
+    props$title <- content@extra$display$title
+  }
+  if (!is.null(content@extra$display$icon)) {
+    props$icon <- content@extra$display$icon
+    icon_deps <- htmltools::findDependencies(props$icon)
+  }
+
+  if (!is.null(tool)) {
+    # Format fails if tool is not present (ellmer v0.3.0, tidyverse/ellmer#691)
+    props$request_call <- format(content@request, show = "call")
+
+    props$title <- props$title %||% tool@annotations$title
+    props$icon <- props$icon %||% tool@annotations$icon
+    icon_deps <- icon_deps %||% htmltools::findDependencies(props$icon)
+  } else {
+    props$request_call <- jsonlite::toJSON(
+      list(
+        id = content@request@id,
+        name = content@request@name,
+        arguments = content@request@arguments
+      ),
+      auto_unbox = TRUE,
+      pretty = 2
+    )
+  }
+
+  props <- list2(!!!props, !!!tool_result_display(content))
+  names(props) <- gsub("_", "-", names(props))
+
+  # Get dependencies for the value because we store its HTML as an attribute
+  # where `processDeps()` won't find it.
+  deps <- htmltools::findDependencies(props$value)
+
+  htmltools::tagList(
+    htmltools::tag("shiny-tool-result", props),
+    deps,
+    icon_deps,
```

Comment:
This way you could reduce the number of objects in scope

```suggestion
    htmltools::findDependencies(props$value),
    htmltools::findDependencies(props$icon),
```